### PR TITLE
Azure OpenAI: migrate to the new API, from "function" to "tools"

### LINKED ISSUE #521: [FEATURE] Azure OpenAI: use tools instead of deprecated functions
image
image
https://learn.microsoft.com/en-au/azure/ai-services/openai/how-to/function-calling?tabs=python#using-function-in-the-chat-completions-api-deprecated

### PR DESCRIPTION
**Pull request by jdubois**
As discussed in #521 there's a new API to call functions, which is documented at https://platform.openai.com/docs/api-reference/chat/create

I'm not sure we should merge it now:

- Those APIs are not compatible, so you use one or the other
- Contrary to the opened ticket, I can currently use the older "function" API call without any issue, not even a warning
- For GPT 4, you need to use the model version "1106-preview", as with the current default one ("0613"), it will fail as this new API is not available.

Fix #521

### CODE REVIEW

File: langchain4j-parent/pom.xml
Comment (Martin7-1) [inline langchain4j-parent/pom.xml:301]: Should add `<scope>import</scope>`

Comment (jdubois) [reply langchain4j-parent/pom.xml:301]: Oh sorry that comes from #523 which should be merged before

**Comment by langchain4j** [comment]
@jdubois That's strange 🤔 Vanilla OpenAI supports both old functions and new tools at the same time, I have just checked on these models:
```
[1] gpt-3.5-turbo
[2] gpt-3.5-turbo-0613
[3] gpt-3.5-turbo-1106
[4] gpt-3.5-turbo-16k
[5] gpt-3.5-turbo-16k-0613
[6] gpt-4
[7] gpt-4-0613
[8] gpt-4-1106-preview
```

**Comment by jdubois** [comment]
@langchain4j when I tested, the old API ("functions") worked everywhere, and the newer one ("tools") only worked on the newer model. Then I don't think the API version is related to the model version, so you could have an older model with a newer API.
BTW did you get your API key to do the integration tests yourself? I'm the (internal) thread trying to solve it, I'm just not sure it's all good you on your side.

**Comment by langchain4j** [comment]
@jdubois I did not get a key yet. I have asked via multiple channels, but no resposne so far :(
Would appreciate a lot if you could facilityte this somehow!

**Comment by langchain4j** [comment]
@jdubois so what do we do with functions/tools then? As far as I understand, we cannot merge this PR now because this will break for user using older models, right? Are there any plans to support both at the same time to ease the transition?

**Comment by jdubois** [comment]
@langchain4j for merging this PR: let me test again, typically with `gpt-4-0613`. If it works with older models, then it will be all good.
For the key, let me ping you privately, you should have had something.

**Comment by jdubois** [comment]
@langchain4j I just did some tests with Azure OpenAI, and:
- Indeed the "old" API works on GPT 3.4 and GPT 4
- I'm now getting some Null responses when testing this PR, and this definitely wasn't the case when I coded it last week

I believe we can stay with the old API: it's "deprecated" but still working everywhere, when the newer one doesn't seem to be very stable at the moment.
We should rather wait a bit before merging this. At least let's wait until you get your API key and can do automated integration testing (which should be OK today!!!).

**Comment by langchain4j** [comment]
@jdubois ok then this will not go into the release next week. I will take a look into this after the release

**Comment by jdubois** [comment]
Argh sorry @langchain4j I tried to rebase with my IDE and I don't know what went wrong, but that's all broken now :-(

**Comment by jdubois** [comment]
Fixed and synced with the main branch :-)

**Comment by langchain4j** [comment]
@jdubois I have noticed that this is not a draft any more. Is it ready for review/merge?

**Comment by jdubois** [comment]
Sorry @langchain4j I'm converting this to "draft" again.
The streaming part doesn't work properly, I need to rework this part completely.

**Comment by langchain4j** [comment]
@jdubois just a gentle reminder :) Are we ready to merge this?

**Comment by jdubois** [comment]
@langchain4j sorry, I need more time to test this.

**Comment by langchain4j** [comment]
@jdubois no rush!

**Comment by jdubois** [comment]
@langchain4j I did more tests so I'm putting this as "ready to review", the only thing that I don't understand here is that when I execute a tool the finish reason is `STOP` and not `TOOL_EXECUTION`.
This would be the logical finish reason, and that's how it's working in the `langchain4j-openai` module, so it seems specific to Azure OpenAI.

**Comment by langchain4j** [comment]
@jdubois there are conflicts now, could you please take a look?

Any idea when funcitons will not be supported (how much time do we have)? Was not able to find it in the official docu...

**Comment by jdubois** [comment]
@langchain4j I fixed the conflicts, and all my tests with GPT 3.5 and GPT 4 are green, so I believe we should merge it now.

File: langchain4j-azure-open-ai/src/main/java/dev/langchain4j/model/azure/AzureOpenAiStreamingResponseBuilder.java
Comment (langchain4j) [inline langchain4j-azure-open-ai/src/main/java/dev/langchain4j/model/azure/AzureOpenAiStreamingResponseBuilder.java]: It takes only the first tool call, is this intentional? Parallel tool calls should already be supported, according to the [docs](https://learn.microsoft.com/en-us/azure/ai-services/openai/how-to/function-calling?tabs=python#supported-models)

Comment (jdubois) [reply langchain4j-azure-open-ai/src/main/java/dev/langchain4j/model/azure/AzureOpenAiStreamingResponseBuilder.java]: It's coming from the OpenAI implementation, but indeed that's wrong.
I'm on it, it'll take some time as I'm adding some integration tests.

Comment (jdubois) [reply langchain4j-azure-open-ai/src/main/java/dev/langchain4j/model/azure/AzureOpenAiStreamingResponseBuilder.java]: I can now have multiple tools calls, but then the answers are streamed synchronously. I'm adding a full test suite, so it's easier to see.

Comment (jdubois) [reply langchain4j-azure-open-ai/src/main/java/dev/langchain4j/model/azure/AzureOpenAiStreamingResponseBuilder.java]: @langchain4j here is the test suite, with 3 calls!

Comment (langchain4j) [reply langchain4j-azure-open-ai/src/main/java/dev/langchain4j/model/azure/AzureOpenAiStreamingResponseBuilder.java]: @jdubois awesome!

Could you please copy&adapt this tests into `AzureOpenAiChatModelIT` for sync model?

There are probably still 2 problems in `InternalAzureOpenAiHelper.toOpenAiMessage()`:
1. `chatRequestAssistantMessage.setFunctionCall(functionCallFrom(message));`: shouldn't `setToolCalls()` be used instead?
2. `return new ChatRequestFunctionMessage(nameFrom(message), toolExecutionResultMessage.text());`: shouldn't `ChatRequestToolMessage` be used instead?

Comment (jdubois) [reply langchain4j-azure-open-ai/src/main/java/dev/langchain4j/model/azure/AzureOpenAiStreamingResponseBuilder.java]: Indeed, I've found several issues there. I unfortunately don't have enough time right now, this is going to take me at least 2 weeks to fix.

Comment (langchain4j) [reply langchain4j-azure-open-ai/src/main/java/dev/langchain4j/model/azure/AzureOpenAiStreamingResponseBuilder.java]: @jdubois no worries! It does not seem to be an urgent change, right?

Comment (jdubois) [reply langchain4j-azure-open-ai/src/main/java/dev/langchain4j/model/azure/AzureOpenAiStreamingResponseBuilder.java]: Should be good now!

**Comment by jdubois** [comment]
Hi @langchain4j can you do another review? This looks OK to me now, and this is now becoming important

**Comment by langchain4j** [comment]
@jdubois ok, will review it asap

**Comment by jdubois** [comment]
@langchain4j can you add @alzimmermsft as a reviewer to this PR? He's working on the Azure OpenAI Java SDK, and he just agreed to do a review, so that should be a great help!

File: langchain4j-azure-open-ai/src/main/java/dev/langchain4j/model/azure/AzureOpenAiStreamingResponseBuilder.java
Comment (alzimmermsft) [inline langchain4j-azure-open-ai/src/main/java/dev/langchain4j/model/azure/AzureOpenAiStreamingResponseBuilder.java:82]: Just so I understand the expected behavior here more, as this is the first time I'm looking at this package, the expected behavior for a `ChatCompletionsToolCall` if it doesn't have an ID set is to use the last seen `ToolExecutionRequestBuilder`?

I have no strong opinion or knowledge on whether that is the correct behavior but wanted to ask more in case that isn't the right behavior.

**Comment by langchain4j** [comment]
@jdubois of course!

**Comment by langchain4j** [comment]
@jdubois but I can't add @alzimmermsft, there is no such user in the reviewers search :/

**Comment by jdubois** [comment]
Oh my fault, I forgot: to be a reviewer, you need to be part of the project.
@langchain4j you can create specific teams for this, I can have a look at what we did on @jhipster if you want.

File: langchain4j-azure-open-ai/src/main/java/dev/langchain4j/model/azure/AzureOpenAiStreamingResponseBuilder.java
Comment (alzimmermsft) [inline langchain4j-azure-open-ai/src/main/java/dev/langchain4j/model/azure/AzureOpenAiStreamingResponseBuilder.java:198]: Can ignore if wanted, is the expectation for these to be used in a threaded scenario? If not, `StringBuilder` is preferable as it doesn't have synchronization as it's not thread safe.

File: langchain4j-azure-open-ai/pom.xml
Comment (langchain4j) [inline langchain4j-azure-open-ai/pom.xml]: nit: this seems to be not required

File: langchain4j-azure-open-ai/src/main/java/dev/langchain4j/model/azure/AzureOpenAiStreamingResponseBuilder.java
Comment (langchain4j) [inline langchain4j-azure-open-ai/src/main/java/dev/langchain4j/model/azure/AzureOpenAiStreamingResponseBuilder.java]: nit: is this required?

File: langchain4j-azure-open-ai/src/main/java/dev/langchain4j/model/azure/InternalAzureOpenAiHelper.java
Comment (langchain4j) [inline langchain4j-azure-open-ai/src/main/java/dev/langchain4j/model/azure/InternalAzureOpenAiHelper.java]: I am not sure I understand how this works.
Here `toolThatMustBeExecuted.name()` (e.g. "calculator", "weather", etc.) is compared with `ChatCompletionsToolSelectionPreset.values()`  (e.g. "auto" or "none").
So, this condition will most likely never be true.

Is this intended logic?

File: langchain4j-azure-open-ai/src/main/java/dev/langchain4j/model/azure/InternalAzureOpenAiHelper.java
Comment (langchain4j) [inline langchain4j-azure-open-ai/src/main/java/dev/langchain4j/model/azure/InternalAzureOpenAiHelper.java]: Is it correct to set empty (`""`) arguments?

File: langchain4j-azure-open-ai/src/test/java/dev/langchain4j/model/azure/AzureOpenAiStreamingChatModelIT.java
Comment (langchain4j) [inline langchain4j-azure-open-ai/src/test/java/dev/langchain4j/model/azure/AzureOpenAiStreamingChatModelIT.java:368]: Could you please add a similar test to the `AzureOpenAiChatModelIT`, to make sure that multiple tools at once can be called with sync model as well?

**Review by ghost — COMMENTED** [review]
@jdubois great job!

File: langchain4j-azure-open-ai/pom.xml
Comment (jdubois) [reply langchain4j-azure-open-ai/pom.xml]: Indeed, it's removed!

File: langchain4j-azure-open-ai/src/main/java/dev/langchain4j/model/azure/AzureOpenAiStreamingResponseBuilder.java
Comment (jdubois) [reply langchain4j-azure-open-ai/src/main/java/dev/langchain4j/model/azure/AzureOpenAiStreamingResponseBuilder.java]: I removed it

File: langchain4j-azure-open-ai/src/main/java/dev/langchain4j/model/azure/InternalAzureOpenAiHelper.java
Comment (jdubois) [reply langchain4j-azure-open-ai/src/main/java/dev/langchain4j/model/azure/InternalAzureOpenAiHelper.java]: Indeed, I removed it

File: langchain4j-azure-open-ai/src/main/java/dev/langchain4j/model/azure/InternalAzureOpenAiHelper.java
Comment (jdubois) [reply langchain4j-azure-open-ai/src/main/java/dev/langchain4j/model/azure/InternalAzureOpenAiHelper.java]: I'm now using `toOpenAiParameters(toolThatMustBeExecuted.parameters()).toString()` to get the arguments. So it's the same as for `setTools`, just we put the parameters into the binary data, unwrap them (to get the JSON), and then wrap them again in the final `setToolChoice`.

File: langchain4j-azure-open-ai/src/test/java/dev/langchain4j/model/azure/AzureOpenAiStreamingChatModelIT.java
Comment (jdubois) [reply langchain4j-azure-open-ai/src/test/java/dev/langchain4j/model/azure/AzureOpenAiStreamingChatModelIT.java:368]: Done!

**Comment by jdubois** [comment]
@langchain4j thanks for the review, this should be all good!

**Review by ghost — APPROVED** [review]
@jdubois thank you!

**Pull request merged**
